ast/compile_test.go

```diff
@@ -875,6 +875,166 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+	}{
+		{
+			note:          "allOf with mergeable Object types in schema",
+			schema:        allOfObjectSchema,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Array types in schema",
+			schema:        allOfArraySchema,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf without a parent schema",
+			schema:        allOfSchemaParentVariation,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with empty schema",
+			schema:        emptySchema,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Array of Object types in schema",
+			schema:        allOfArrayOfObjects,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Object types in schema with type declaration missing",
+			schema:        allOfObjectMissing,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with Array of mergeable different types in schema",
+			schema:        allOfArrayDifTypes,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Object containing Array types in schema",
+			schema:        allOfArrayInsideObject,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Array types in schema with type declaration missing",
+			schema:        allOfArrayMissing,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable types inside of core schema",
+			schema:        allOfInsideCoreSchema,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable String types in schema",
+			schema:        allOfStringSchema,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Integer types in schema",
+			schema:        allOfIntegerSchema,
+			expectedError: nil,
+		},
+		{
+			note:          "allOf with mergeable Boolean types in schema",
+			schema:        allOfBooleanSchema,
+			expectedError: nil,
+		},
+		{
+			note:          "schema of array type with uneven numbers of items to merge under allOf",
+			schema:        allOfSchemaWithUnevenArray,
+			expectedError: nil,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.note, func(t *testing.T) {
+			if tc.expectedError != nil {
+				c := NewCompiler()
+				var schema interface{}
+				err := util.Unmarshal([]byte(tc.schema), &schema)
+				if err != nil && err != tc.expectedError {
+					t.Fatal("Unexpected error:", err)
+				}
+				schemaSet := NewSchemaSet()
+				schemaSet.Put(SchemaRootRef, schema)
+				c.WithSchemas(schemaSet)
+				compileStages(c, c.checkTypes)
+				expected := []string{}
+				expected = append(expected, tc.expectedError.Error())
+				assertCompilerErrorStrings(t, c, expected)
+			} else {
+				c := NewCompiler()
+				var schema interface{}
+				err := util.Unmarshal([]byte(tc.schema), &schema)
+				if err != nil {
+					t.Fatal("Unexpected error:", err)
+				}
+				schemaSet := NewSchemaSet()
+				schemaSet.Put(SchemaRootRef, schema)
+				c.WithSchemas(schemaSet)
+				compileStages(c, c.checkTypes)
```

Comment:
You dont need to repeat all these lines of code again inside the `else`. To minimize code repetition, all you need is the `if tc.expectedError != nil {` right before the assert statements. Not the whole code...